internal/proxy/proxy.go

```diff
@@ -0,0 +1,84 @@
+package proxy
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/buzzfeed/sso/internal/pkg/hostmux"
+	"github.com/buzzfeed/sso/internal/pkg/options"
+	"github.com/buzzfeed/sso/internal/pkg/ping"
+)
+
+type SSOProxy struct {
+	http.Handler
+}
+
+func New(opts *Options) (*SSOProxy, error) {
+	optFuncs := []func(*OAuthProxy) error{}
+
+	var requestSigner *RequestSigner
+	var err error
+	if opts.RequestSigningKey != "" {
+		requestSigner, err = NewRequestSigner(opts.RequestSigningKey)
+		if err != nil {
+			return nil, err
+		}
+		optFuncs = append(optFuncs, SetRequestSigner(requestSigner))
+	}
+
+	if len(opts.EmailAddresses) != 0 {
+		optFuncs = append(optFuncs, SetValidator(options.NewEmailAddressValidator(opts.EmailAddresses)))
+	} else {
+		optFuncs = append(optFuncs, SetValidator(options.NewEmailDomainValidator(opts.EmailDomains)))
+	}
+
+	hostRouter := hostmux.NewRouter()
+	for _, upstreamConfig := range opts.upstreamConfigs {
+		var handler http.Handler
+		switch route := upstreamConfig.Route.(type) {
+		case *SimpleRoute:
+			reverseProxy := NewReverseProxy(route.ToURL, upstreamConfig)
+			handler, _ = NewReverseProxyHandler(
+				reverseProxy, opts, upstreamConfig, requestSigner)
+		case *RewriteRoute:
+			reverseProxy := NewRewriteReverseProxy(route, upstreamConfig)
+			handler, _ = NewReverseProxyHandler(
+				reverseProxy, opts, upstreamConfig, requestSigner)
+		default:
+			return nil, fmt.Errorf("unknown route type")
+		}
+
+		optFuncs = append(optFuncs,
+			SetCookieStore(opts),
+			SetUpstreamConfig(upstreamConfig),
+			SetProxyHandler(handler),
```

Comment:
`SetProxyHandler` strikes me as a bad use of functional options, because a handler is not an _optional_ piece of an `OAuthProxy` instance.  My vote would be to change the signature of `NewOAuthProxy` to take a handler as its first argument:

```go
oauthproxy, err := NewOAuthProxy(handler, opts, optFuncs...)
```

Reply:
It's particularly nice for testing though :)

Reply:
Do we have a lot of tests where we don't need a handler set up?  Can we paper over that in a testing helper function?

Reply:
https://github.com/buzzfeed/sso/pull/190/commits/b930cc51d996092d80971ab427245e0820e3695c

I've added this commit to add error messages if the `NewOAuthProxy` does is not properly configured. 

I think this is a nice compromise between the func options as well as ensuring the oauthproxy we return is functional.